fix(compaction group): check existence of given compaction group id b…

soundOfDestiny opened the pull request:
…efore hummock manager operations

I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- check existence of given compaction group id before hummock manager operations, this becomes necessary because it is possble that certain compaction group has been deleted now
- let `HummockVersion` and `CompactionGroup` in meta be consistent with `CompactionGroupManager`
https://singularity-data.quip.com/iohfAKFRfxtu/Consistent-among-CompactionGroupManager-HummockVersion-and-CompactStatus-in-meta

## Checklist

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

zwang28 commented:
Please don't review for now.
Need more discussion in terms of metadata and SST deletion.

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -1149,10 +1167,72 @@ where
+                    delta_type: DeltaType::Groupdestroy as i32,
+                    level_idx: u32::MAX,
+                    l0_sub_level_id: u64::MAX,
+                    ..Default::default()
```
Should also set removed_table_ids field, so that group's SSTs can be removed in extend_ssts_to_delete_from_deltas.

soundOfDestiny replied:
> Should also set removed_table_ids field, so that group's SSTs can be removed in extend_ssts_to_delete_from_deltas.

No. Currently we remove after SSTs are deleted.

soundOfDestiny replied:
> Should also set removed_table_ids field, so that group's SSTs can be removed in extend_ssts_to_delete_from_deltas.

fixed

zwang28 commented on proto/hummock.proto:
```diff
@@ -40,10 +40,17 @@ message InputLevel {
+  uint64 l0_sub_level_id = 3;
+  repeated uint64 removed_table_ids = 4;
+  repeated SstableInfo inserted_table_infos = 5;
+  CompactionConfig group_config = 6;
```
group_config is only used by GROUPCONSTRUCT.
And if we want to make a LevelDelta self-contained, GROUPDESTROY need know group id.
May consider oneof to carry data for DeltaType?

soundOfDestiny replied:
> oneof

fixed

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -1149,10 +1167,72 @@ where
+                new_hummock_version
+                    .levels
+                    .try_insert(group_id, {
+                        let mut levels = vec![];
```
Consider extract levels initialization, as we have some duplicate now.

soundOfDestiny replied:
> Consider extract levels initialization, as we have some duplicate now.

fixed

zwang28 commented on src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:
```diff
@@ -450,11 +302,206 @@ impl HummockVersionDeltaExt for HummockVersionDelta {
             .values()
             .flat_map(|item| item.level_deltas.iter())
-            .map(|level_delta| level_delta.inserted_table_infos.len())
+            .map(|level_delta| match level_delta.get_delta_type().unwrap() {
```
`get_sstableinfo_cnt` is not used at all, and its meaning is not clear. Can we remove it from HummockVersionDeltaExt trait?

soundOfDestiny replied:
> `get_sstableinfo_cnt` is not used at all, and its meaning is not clear. Can we remove it from HummockVersionDeltaExt trait?

removed

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -607,15 +595,15 @@ where
+            .ok_or(Error::InvalidCompactionGroup(compaction_group_id))?;
         if !compaction
             .compaction_statuses
             .contains_key(&compaction_group_id)
```
Can we add comment/doc about lifecycle of both compaction group entry in CompactStatus and HummockVersion, like when they are (lazily)initialized/destroyed, when it's an expected behavior or a fatal bug if trying to visit a missing group entry?

soundOfDestiny replied:
> Can we add comment/doc about lifecycle of both compaction group entry in CompactStatus and HummockVersion, like when they are (lazily)initialized/destroyed, when it's an expected behavior or a fatal bug if trying to visit a missing group entry?

https://singularity-data.quip.com/iohfAKFRfxtu/Consistent-among-CompactionGroupManager-HummockVersion-and-CompactStatus-in-meta

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -901,14 +892,25 @@ where
-        );
+        let compaction_groups: HashSet<_> = self
+            .compaction_group_manager
+            .compaction_groups()
```
Clone entire compaction group can be expensive because it may contain many table_ids. This is an existent issue before this PR. Now that we have quite some usage of compaction_groups, let's consider optimize this (maybe in another PR).

soundOfDestiny replied:
> existent

I'm sorry. We can call `compaction_group_ids()` here instead. fixed.

zwang28 reviewed (commented):
LGTM. 
Please update the PR description.
And let's add tests and docs.

soundOfDestiny commented:
> LGTM.
> Please update the PR description.
> And let's add tests and docs.

how to add

soundOfDestiny commented:
> LGTM.
> Please update the PR description.
> And let's add tests and docs.

tests added

soundOfDestiny commented:
> …efore hummock manager operations
> 
> I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).
> 
> ## What's changed and what's your intention?
> check existence of given compaction group id before hummock manager operations
> 
> this becomes necessary because it is possble that certain compaction group has been deleted now
> 
> ## Checklist
> ## Documentation
> If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.
> 
> ### Types of user-facing changes
> Please keep the types that apply to your changes, and remove those that do not apply.
> 
> * Installation and deployment
> * Connector (sources & sinks)
> * SQL commands, functions, and operators
> * RisingWave cluster configuration changes
> * Other (please specify in the release note below)
> 
> ### Release note
> Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.
> 
> ## Refer to a related PR or issue link (optional)



> LGTM.
> Please update the PR description.
> And let's add tests and docs.

docs added.
PR description updated

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -901,14 +892,24 @@ where
+        let mut compact_statuses = BTreeMapTransaction::new(&mut compaction.compaction_statuses);
+        for group_id in original_keys {
+            if !compaction_groups.contains(&group_id) {
+                compact_statuses.remove(group_id);
```
Directly remove compact_status may fail compaction heartbeat cancellation
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L249

which cleans up compaction heartbeat in
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L989

So when removing compact_status, we need also consider compact_task_assignment and compaction heartbeat.

zwang28 replied:
Similar here
https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/context.rs#L52

This call will be retried until success. But it can fail forever due to group's compact_status absence.

soundOfDestiny replied:
> Directly remove compact_status may fail compaction heartbeat cancellation
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L249
> 
> which cleans up compaction heartbeat in
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/mod.rs#L989
> 
> So when removing compact_status, we need also consider compact_task_assignment and compaction heartbeat.

fixed

soundOfDestiny replied:
> Similar here
> 
> https://github.com/risingwavelabs/risingwave/blob/cf5852f248735095f54f2b971153fbf4f0ba2a32/src/meta/src/hummock/manager/context.rs#L52
> 
> This call will be retried until success. But it can fail forever due to group's compact_status absence.

fixed

zwang28 commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -940,7 +930,16 @@ where
+                compact_status.report_compact_task(compact_task);
+            }
+            None => {
+                compact_task.set_task_status(TaskStatus::HeartbeatCanceled);
```
Use a new accurate TaskStatus?

soundOfDestiny replied:
> Use a new accurate TaskStatus?

fixed

zwang28 reviewed (commented):
LGTM.
Before merge, I suggest we 

- replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
- provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
- add tests that verifies correctness of compaction group adding/removing.

soundOfDestiny commented:
> LGTM. Before merge, I suggest we
> 
> * replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
> * provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
> * add tests that verifies correctness of compaction group adding/removing.

![127 0 0 1_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s](https://user-images.githubusercontent.com/44948473/192733230-266c1a21-3d3b-4586-8264-b5e5c40bc319.png)

soundOfDestiny commented:
> LGTM. Before merge, I suggest we
> 
> * replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
> * provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
> * add tests that verifies correctness of compaction group adding/removing.

main:
![127 0 0 1_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s (1)](https://user-images.githubusercontent.com/44948473/192745674-b8f940cb-f892-4d1d-b809-9ef61976a67f.png)

soundOfDestiny commented:
> LGTM. Before merge, I suggest we
> 
> * replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
> * provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
> * add tests that verifies correctness of compaction group adding/removing.

task 2 finished
task 1 finished
task 3 finished by task 1

soundOfDestiny commented:
bench sql:
```
create source person
    (id INTEGER, name VARCHAR, "email_address" VARCHAR, "credit_card" VARCHAR, city VARCHAR, state VARCHAR, "date_time" TIMESTAMP)
with (
  connector = 'nexmark',
  nexmark.table.type = 'Person',
  nexmark.split.num = '24',
    nexmark.min.event.gap.in.ns = '0'
) row format json;

create source auction (id INTEGER, "item_name" VARCHAR, description VARCHAR, "initial_bid" INTEGER, reserve INTEGER, "date_time" TIMESTAMP, expires TIMESTAMP, seller INTEGER, category INTEGER) 
with (
  connector = 'nexmark',
  nexmark.table.type = 'Auction',
  nexmark.split.num = '24',
  nexmark.min.event.gap.in.ns = '0'
) row format json;

create source bid (auction INTEGER, bidder INTEGER, price INTEGER, "date_time" TIMESTAMP)
with (
  connector = 'nexmark',
  nexmark.table.type = 'Bid',
  nexmark.split.num = '24',
  nexmark.min.event.gap.in.ns = '0'
) row format json;

CREATE MATERIALIZED VIEW nexmark_q7 AS
SELECT
      B.auction,
      B.price,
      B.bidder,
      B.date_time
    FROM
      bid B
    JOIN (
      SELECT
        MAX(price) AS maxprice,
        window_end as date_time
      FROM
        TUMBLE(bid, date_time, INTERVAL '10' SECOND)
      GROUP BY
        window_end
    ) B1 ON B.price = B1.maxprice
    WHERE
      B.date_time BETWEEN B1.date_time - INTERVAL '10' SECOND
      AND B1.date_time;
```

soundOfDestiny commented:
> LGTM. Before merge, I suggest we
> 
> * replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
> * provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
> * add tests that verifies correctness of compaction group adding/removing.

tpch result, in which source throughput > 35000
![127 0 0 1_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s (2)](https://user-images.githubusercontent.com/44948473/193025213-d3cbafd8-a708-462b-a0db-5a3b4f3f057d.png)

soundOfDestiny commented:
> LGTM. Before merge, I suggest we
> 
> * replace https://github.com/risingwavelabs/risingwave/blob/4998ba94da6f85f1568fa36a399e21a5797b05b9/src/frontend/src/handler/create_mv.rs#L122
> * provide benchmark result to compare performance of main's 2 static group with this PR's more dedicated groups.
> * add tests that verifies correctness of compaction group adding/removing.

tech result main:
![127 0 0 1_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s from=now-1h to=now (1)](https://user-images.githubusercontent.com/44948473/193027473-bc634c36-941b-489d-95d3-7a9632d277dd.png)

soundOfDestiny commented:
@zwang28 Could we finish it? I plan to issue an optimization scheme for compaction groups today.

zwang28 commented on src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:
```diff
@@ -275,7 +304,7 @@ impl HummockLevelsExt for Levels {
             }
         }
-        if local_related_only && !deleted {
+        if local_related_only && !delete_sst_ids_set.is_empty() && !deleted {
```
Could you explain why !delete_sst_ids_set.is_empty()?

soundOfDestiny replied:
> Could you explain why !delete_sst_ids_set.is_empty()?

logic of codes here is that: 
- (a) This is a compaction
- (b) generated SSTs are from `delete_sst_ids_set`
so if we pay no attention to SSTs in `delete_sst_ids_set`, we shall also pay no attention to generated SSTs

However, if `delete_sst_ids_set` is empty, Condition (b) is not maintained

soundOfDestiny replied:
> Could you explain why !delete_sst_ids_set.is_empty()?

> Could you explain why !delete_sst_ids_set.is_empty()?

logic of codes here is that: 
- (a) This is a compaction
- (b) generated SSTs are from `delete_sst_ids_set`

Therefore if we pay no attention to SSTs in `delete_sst_ids_set`, we shall also pay no attention to generated SSTs

However, if `delete_sst_ids_set` is empty, Condition (b) is not maintained

zwang28 commented on src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:
```diff
@@ -362,6 +411,9 @@ pub fn add_new_sub_level(
     insert_sub_level_id: u64,
     insert_table_infos: Vec<SstableInfo>,
 ) {
+    if insert_sub_level_id == u64::MAX {
```
When will this be true?

soundOfDestiny replied:
> When will this be true?

When summary doesn't find any insert operations in `level_delta`

zwang28 commented on proto/hummock.proto:
```diff
@@ -41,13 +41,27 @@ message InputLevel {
+  CompactionConfig group_config = 1;
+}
+
+message GroupDestroy {}
```
What about keep group id in GroupDestroy?

soundOfDestiny replied:
> What about keep group id in GroupDestroy?

Currently not, because `GroupDestroy` is in `level_deltas`.
In the future, if we have `GroupSplit`/`GroupMerge` directly in `VersionDelta`(not `level_deltas`), we can then move `GroupDestroy` there and then keep group id in `GroupDestroy`

zwang28 commented on proto/hummock.proto:
```diff
@@ -75,11 +89,11 @@ message HummockVersionDelta {
   // Levels of each compaction group
   map<uint64, LevelDeltas> level_deltas = 3;
-  uint64 max_committed_epoch = 4;
+  uint64 max_committed_epoch = 5;
```
We shouldn't change original indexes. Same below.

soundOfDestiny replied:
> We shouldn't change original indexes. Same below.

fixed

zwang28 commented on proto/hummock.proto:
```diff
@@ -187,9 +201,10 @@ message CompactTask {
-    EXECUTE_FAILED = 8;
-    JOIN_HANDLE_FAILED = 9;
-    TRACK_SST_ID_FAILED = 10;
+    INVALIDGROUP_CANCELED = 8;
```
Please use new index for new enum. We shouldn't change original indexes.

soundOfDestiny replied:
> Please use new index for new enum. We shouldn't change original indexes.

I want to arrange `CANCELED` together and arrange `FAILED` together, respectively.

soundOfDestiny commented on src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:
```diff
@@ -39,14 +41,16 @@ pub fn summarize_level_deltas(level_deltas: &LevelDeltas) -> LevelDeltasSummary
-            insert_sst_level_id = level_delta.level_idx;
-            insert_sub_level_id = level_delta.l0_sub_level_id;
-            insert_table_infos.extend(level_delta.inserted_table_infos.iter().cloned());
+        if let DeltaType::IntraLevel(intra_level) = level_delta.get_delta_type().unwrap() {
```
> Shall we add GroupConstruct and GroupDestroy in LevelDeltasSummary? So that someone is less likely to appy IntraLevel without constructing groups beforehand and destroying groups afterwards.

fixed

zwang28 reviewed (approved):
LGTM. @Little-Wallace @hzxa21 @Li0k PTAL

- This PR fixes issues of runtime compaction group creation/removal.
- CREATE_COMPACTION_GROUP_FOR_MV is false by default, which means we still only have 2 static compaction groups by default. I've benched with nextmark_q7, which shows turning it on causes no significant changes in terms of key metrics. I think we can bench with more workload later, and more integration tests as well(working on it).
- Will cause some conflict in https://github.com/risingwavelabs/risingwave/pull/5441, e.g. delta log now have 2 new types named GroupConstruct and GroupDestroy. I think we can keep the assumption of 2 static compaction group in the first version of compaction deterministic test tool, if it requires lots work to adapt to arbitrary number of compaction group. @StrikeW

soundOfDestiny commented:
> * compaction group in the first version of compaction deterministic test tool, if it requires lots work to adapt to arbitrar

@Gun9niR

Li0k commented:
I will review it ASAP

StrikeW commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -1166,10 +1172,82 @@ where
         new_version_delta.id = new_version_id;
+        new_hummock_version.id = new_version_id;
 
+        for group_id in old_version_groups {
```
It would be better to split these codes to another function and have some comments on it to make it easy to read.

soundOfDestiny replied:
> It would be better to split these codes to another function and have some comments on it to make it easy to read.

ok. let me refactor.

Gun9niR commented:
> > * compaction group in the first version of compaction deterministic test tool, if it requires lots work to adapt to arbitrar
> 
> @Gun9niR @wenym1

Why do we need this assumption?

StrikeW commented:
> > > I think we can keep the assumption of 2 static compaction group in the first version of compaction deterministic test tool, if it requires lots work to adapt to arbitrar
> > 
> > 
> > @Gun9niR @wenym1
> 
> Why do we need this assumption?

The test tool will get the modified compaction groups from HummockVersionDelta, currently via getting the keys from field `level_deltas`. I think it doesn't need to make assumption of the number of compaction groups.

soundOfDestiny commented on proto/hummock.proto:
```diff
@@ -75,11 +89,11 @@ message HummockVersionDelta {
   // Levels of each compaction group
   map<uint64, LevelDeltas> level_deltas = 3;
-  uint64 max_committed_epoch = 4;
+  uint64 max_committed_epoch = 5;
```
> mistake to modify the tag？

sorry. fixed.

Li0k commented on src/frontend/src/handler/create_mv.rs:
```diff
@@ -114,18 +114,26 @@ pub async fn handle_create_mv(
+            context.into(),
+            query,
+            name,
+            is_independent_compaction_group,
```
function `gen_create_mv_plan` can get the `session`, can we remove the variable `is_independent_compaction_group`, and inline this logic

soundOfDestiny replied:
> function `gen_create_mv_plan` can get the `session`, can we remove the variable `is_independent_compaction_group`, and inline this logic

thank you. it's a mistake.

Li0k commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -178,6 +182,7 @@ static CANCEL_STATUS_SET: LazyLock<HashSet<TaskStatus>> = LazyLock::new(|| {
         TaskStatus::SendFailCanceled,
         TaskStatus::AssignFailCanceled,
         TaskStatus::HeartbeatCanceled,
+        TaskStatus::InvalidgroupCanceled,
```
`InvalidgroupCanceled` -> InvalidGroupCanceled

soundOfDestiny replied:
> `InvalidgroupCanceled` -> InvalidGroupCanceled

fixed

Li0k commented on src/meta/src/hummock/manager/mod.rs:
```diff
@@ -1113,6 +1109,93 @@ where
+        // Insert new groups that are not in current `HummockVersion`;
+        // Delete old groups that still remain in current `HummockVersion`.
+
+        // Delete old groups that still remain in current `HummockVersion`.
```
dup notes ？

soundOfDestiny replied:
> dup notes ？

deleted one

hzxa21 commented on proto/hummock.proto:
```diff
@@ -41,13 +41,27 @@ message InputLevel {
+
+message GroupDestroy {}
+
+message LevelDelta {
```
Naming it `LevelDelta` is a bit confusing since `GroupConstruct` and `GrounpDestroy` have nothing to do with level. How about naming it `GroupDelta` or `GroupVersionDelta` instead?

soundOfDestiny replied:
> Naming it `LevelDelta` is a bit confusing since `GroupConstruct` and `GrounpDestroy` have nothing to do with level. How about naming it `GroupDelta` or `GroupVersionDelta` instead?

ok
